keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1537 @@
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Implementation
+
+![components](./components.png)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which generates
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and then invoke local resource plugin methods. It must pass information about
+  the additional resources to the container runtime. It must detect
+  whether the container runtime has the necessary support and
+  advertise that in the cluster via node labels to simplify deployment
+  of plugins and Pod scheduling.
+
+For a resource plugin the following components are needed:
+- Some utility library similar to
+  https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner
+  and the code in driver-registrar.
+- *Resource controller*: a central component which handles resource allocation
+  by watching and modifying ResourceClaims.
+- *Resource node plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+The utility library will be developed outside of Kubernetes and does not have
+to be used by plugins, therefore it is not described further in this KEP.
+
+
+```
+<<[UNRESOLVED @pohly]>>
+All of the changes in Kubernetes need to be specified in a lot more detail.
+
+Upgrade and downgrade scenarios should already be considered for v1alpha1 to ensure
+that whatever changes will be needed are in place before going to v1beta1 where
+downgrades have to be supported.
+
+All of that will be added once there is consensus to move ahead with this proposal.
+<<[/UNRESOLVED]>>
+```
+
+### Resource allocation flow
+
+The following diagram shows how resource allocation works for a resource that
+gets defined inline in a Pod. For a full definition of ResourceClass,
+ResourceClaim and ResourceClaimTemplate see the [API](#API) section below.
+
+Several of these operations may fail without changing the system state. They
+then must be retried until they succeed or something else changes in the
+system, like for example deleting objects. These additional state transitions
+are not shown for the sake of simplicity.
+
+![allocation](./allocation.png)
+
```

Comment:
The ResourceController modifies the set of node candidates that the scheduler uses for scheduling the POD, so the ResourceController seems to be a scheduler plugin registered at a scheduler extension point, right?

Reply:
You refer to "scheduler resource controller" in allocation.puml, right? That's a mistake, it should be "resource claim controller". It's the in-tree part that runs inside kube-controller-manager.

After the discussion in https://github.com/kubernetes/enhancements/pull/3064#issuecomment-1014537786, we should have a "[kube-controller-manager] resource claim controller" and a "[third-party] resource driver controller".  The scheduler plugin should be called "[kube-]scheduler resource plugin".

I missed some mentions of "resource controller" without the clarifications. Should be fixed now.